### PR TITLE
Add the data path filtering for policy verdict logs.

### PR DESCRIPTION
**Pull request by lzang**
Generate policy verdict logs only if the endpoint does have a network policy
enforced on the direction of the traffic.

Signed-off-by: Zang Li <zangli@google.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10477)

### CODE REVIEW

File: bpf/lib/policy_log.h
Comment (aanm) [inline bpf/lib/policy_log.h]: Does this need a follow up?

**Comment by pchaigno** [comment]
test-me-please

File: bpf/lib/policy_log.h
Comment (lzang) [reply bpf/lib/policy_log.h]: It would be nice if someone from Cilium can help. It will also be very useful if we can have a way to load a bpf program and easily test the functionality on different kernel versions.

**Comment by lzang** [comment]
test-me-please

File: pkg/endpoint/bpf.go
Comment (joestringer) [inline pkg/endpoint/bpf.go]: This should only be necessary for the template, so if you added the same function `GetPolicyVerdictLogFilter()` for the `templateCfg` type which set a dummy non-zero value then it should be safe to clear this bit clear for the actual endpoints.

File: bpf/lib/policy_log.h
Comment (joestringer) [reply bpf/lib/policy_log.h]: Are you hitting issues with the current version in this PR?

Which kernel and compiler are you using when you hit these issues?

I would love to have a way to load & test a BPF program against different kernel verifier versions easily, but there's a lot of combinations and that code is unfortunately not particularly easy to rip out and run standalone.

Overall our current CI testing is based on validating that with LLVM-7 and either a recent 4.9.x kernel version or a recent bpf-next / net-next kernel prerelease version, and if those two work, then we can pass through to the next stage which typically involves developers and users running the code on a wider variety of kernels. If this code is not performing correctly on a particular kernel, that may be a kernel bug; in that case, we need to identify whether the bug was already fixed somewhere (eg by testing other kernel versions) and whether the relevant fix was backported by either upstream maintainers or distro maintainers.

Comment (lzang) [reply bpf/lib/policy_log.h]: "Are you hitting issues with the current version in this PR?" Do you mean current code? Current code in this PR works. The two commented out ways didn't work for me. I hit the issue for both 4.14 and 4.19 kernel.

File: pkg/endpoint/bpf.go
Comment (lzang) [reply pkg/endpoint/bpf.go]: Updated.

**Comment by lzang** [comment]
test-me-please

File: pkg/endpoint/cache.go
Comment (joestringer) [inline pkg/endpoint/cache.go]: One more question, if this becomes a uint32 when plumbed to the datapath then shouldn't the rest of the code in userspace assume this size as well? Just to make sure that we don't somehow end up truncating values at some point in the future.

File: bpf/lib/policy_log.h
Comment (pchaigno) [reply bpf/lib/policy_log.h]: I tried the two commented lines on a 4.19 kernel with Clang 7 and all policy verdicts are filtered regardless of network policies. If I try the third line (i.e., this pull request), it works as expected, I see policy verdicts according to ingress/egress network policies. @lzang Is that the issue you were having as well?

Comment (lzang) [reply bpf/lib/policy_log.h]: Yes, that is the issue I had. Thanks for trying it out!

File: pkg/endpoint/cache.go
Comment (lzang) [reply pkg/endpoint/cache.go]: Yes, it is uint32 when plumbed to the datapath. Updated the patch to use uint32 for all places.

**Comment by lzang** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: bpf/lib/policy_log.h
Comment (pchaigno) [reply bpf/lib/policy_log.h]: Ok. So I added a function in its own section in `bpf_lxc.c` with only a call to `send_policy_verdict_notify()` to be able to look at the generated bytecode for the diverse C codes above. With the third line (the one that works), I get:
```
       0:	bf 16 00 00 00 00 00 00	r6 = r1
       1:	18 01 00 00 01 00 00 80 00 00 00 00 00 00 00 00	r1 = 2147483649 ll
       3:	18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00	r2 = 0 ll
       5:	5f 12 00 00 00 00 00 00	r2 &= r1
       6:	15 02 1f 00 00 00 00 00	if r2 == 0 goto +31 <LBB14_4>
[...]
0000000000000130 LBB14_4:
      38:	b7 00 00 00 00 00 00 00	r0 = 0
      39:	95 00 00 00 00 00 00 00	exit
```
Whereas, with the second line, I get:
```
       0:	b7 00 00 00 00 00 00 00	r0 = 0
       1:	95 00 00 00 00 00 00 00	exit
```
So the compiler seems to think that, because `filter` is 0 at compile time, it can optimize out this logic and get rid of most of the `send_policy_verdict_notify` function. It's not yet clear to me why the compiler think it's okay to optimize or why the third line doesn't get optimized in a similar fashion.

In any case, we may want to use `volatile` here instead and add a comment to explain.
```c
volatile __u8 d = dir;
return ((filter & d) > 0);
```
Both Clang 7 and 9 behave this way. ~~I haven't checked newer versions yet.~~ Latest Clang (trunk) does as well.

File: bpf/lib/static_data.h
Comment (pchaigno) [inline bpf/lib/static_data.h]: Is the use of `volatile` required here?

File: bpf/lib/policy_log.h
Comment (pchaigno) [inline bpf/lib/policy_log.h]: Since this argument is the same from all call sites (at least for now) and is a global (config) variable, maybe we could get rid of that argument and instead directly match aginst `POLICY_VERDICT_LOG_FILTER` in `policy_verdict_filter_allow`?

File: bpf/lib/static_data.h
Comment (joestringer) [reply bpf/lib/static_data.h]: The goal of the `volatile` here is to ensure that the compiler does not make any assumptions about the value held by `NAME`, so that at compile time both paths are generated, then at load time we can substitute the appropriate value.

I expect that this is particularly important when the value of `NAME` is used for conditional code execution.

Comment (pchaigno) [reply bpf/lib/static_data.h]: Ah, I see. It seemed weird that above macros didn't have the same, but that makes sense.

**Comment by joestringer** [comment]
From here:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17796
Through here:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/17796/testReport/junit/Suite-k8s-1/11/K8sDatapathConfig_DirectRouting_Check_direct_connectivity_with_per_endpoint_routes/

Into the zip, into the cilium logs for pod `cilium-cllgt`, filtering out all `level=debug` / `level=info`, I found repeats of this set of complaints:

```
2020-03-09T02:54:00.2459225Z level=error msg="Command execution failed" cmd="[tc filter replace dev lxc5ea689f5bbdb ingress prio 1 handle 1 bpf da obj 564_next/bpf_lxc.o sec from-container]" error="exit status 1" subsys=datapath-loader
2020-03-09T02:54:00.245943042Z level=warning msg="Log buffer too small to dump verifier log 16777215 bytes (10 tries)!" subsys=datapath-loader
2020-03-09T02:54:00.24595512Z level=warning msg="Error fetching program/map!" subsys=datapath-loader
2020-03-09T02:54:00.245958089Z level=warning msg="Unable to load program" subsys=datapath-loader
2020-03-09T02:54:00.263696359Z level=warning msg="JoinEP: Failed to load program" containerID=70678a65fa datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=564 error="Failed to load tc filter: exit status 1" file-path=564_next/bpf_lxc.o identity=58084 ipv4=10.10.1.179 ipv6= k8sPodName=default/test-k8s2-8d47f6c76-mpxvg subsys=datapath-loader veth=lxc5ea689f5bbdb
2020-03-09T02:54:00.263715579Z level=error msg="Error while rewriting endpoint BPF program" containerID=70678a65fa datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=564 error="Failed to load tc filter: exit status 1" identity=58084 ipv4=10.10.1.179 ipv6= k8sPodName=default/test-k8s2-8d47f6c76-mpxvg subsys=endpoint
2020-03-09T02:54:00.264195958Z level=warning msg="generating BPF for endpoint failed, keeping stale directory." containerID=70678a65fa datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=564 file-path=564_next_fail identity=58084 ipv4=10.10.1.179 ipv6= k8sPodName=default/test-k8s2-8d47f6c76-mpxvg subsys=endpoint
2020-03-09T02:54:00.264422446Z level=warning msg="Regeneration of endpoint failed" bpfCompilation=1.389299726s bpfLoadProg=13.921122438s bpfWaitForELF=1.389554451s bpfWriteELF="211.137µs" buildDuration=15.316466496s containerID=70678a65fa datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=564 error="Failed to load tc filter: exit status 1" identity=58084 ipv4=10.10.1.179 ipv6= k8sPodName=default/test-k8s2-8d47f6c76-mpxvg mapSync="21.086µs" policyCalculation="162.228µs" prepareBuild="332.159µs" proxyConfiguration="4.986µs" proxyPolicyCalculation="132.499µs" proxyWaitForAck=0s reason="updated security labels" subsys=endpoint waitingForCTClean=2.773349ms waitingForLock=883ns
2020-03-09T02:54:00.264678764Z level=error msg="endpoint regeneration failed" containerID=70678a65fa datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=564 error="Failed to load tc filter: exit status 1" identity=58084 ipv4=10.10.1.179 ipv6= k8sPodName=default/test-k8s2-8d47f6c76-mpxvg subsys=endpoint
```

I'm not sure why the Jenkins CI didn't identify any errors/warnings in the logs, but I saw them there.

This particular case I think we've seen before on the `net-next` kernel image, do you recall the workaround there?

For the other two failures from that link, they look like policy-related CI flakes. Given the timing they may have been resolved by #10493.

At a glance from the latest failures, I see endpoints in `not-ready` state, and I was able to find evidence of the above "log buffer too small" issue again.

**Comment by lzang** [comment]
We bypassed the issue earlier by removing all the if else statements in the datapath patch. For this one, we cannot bypass the single check for the filtering operation.

File: bpf/lib/policy_log.h
Comment (lzang) [reply bpf/lib/policy_log.h]: @pchaigno Thanks for your investigation. We've already made the filter to be always none-zero as we knew initializing it to zero causes issue. Filter is a volatile uint32 and its value is correctly interpreted. The issue here is dir, which is an input to the function and is always non-zero. Making a local variable to be volatile looks strange to me. Did you try this change end-to-end?

Comment (pchaigno) [reply bpf/lib/policy_log.h]: `filter` is actually zero when compiling (`r2` in my above snippet) and is replaced with an actual value when we rewrite the object file. [This LPC presentation](http://vger.kernel.org/lpc_bpf2018_talks/bpf_global_data_and_static_keys.pdf) gives some background and explanations for those config. variables. Basically, `POLICY_VERDICT_LOG_FILTER` ends up taking the address of the global variable and not its value; that allows us to easily rewrite it in the object file afterwards. In BPF, such addresses of global variables are usually references to maps and rewritten in the kernel at load time to point to the actual BPF map. *My guess* is that the compiler sets it to zero because it assumes it's one of this map cases where the compile-time value doesn't matter.

I did test the change I proposed above. I executed Cilium with the same setup as before (Clang 7) and checked that policy verdicts were filtered by default and started showing up when network policies were added. I did not run the whole test suite though.

Comment (lzang) [reply bpf/lib/policy_log.h]: @pchaigno What I don't understand is that you are talking about filter being zero or not, but your fix is on dir. Making dir to be volatile or adding some constant to it are both trying to change dir.

File: bpf/lib/policy_log.h
Comment (lzang) [reply bpf/lib/policy_log.h]: How to do that? Including the header file (lxc_config.h ) will result in the build fail.  What is the alternative way?

File: bpf/lib/policy_log.h
Comment (pchaigno) [reply bpf/lib/policy_log.h]: You're right, that wasn't super clear. There are two different things here:
1. Why the compiler thinks the condition is always equivalent to zero. That's because `filter` is always zero at compile time and that's what I tried to explain in my previous comment.
0. Why making `dir` volatile prevents the code optimization. I don't fully understand that one. I would expect that having one variable volatile in the expression (i.e., `filter`) is enough, but that doesn't seem to be the case. Making both variables volatile seems to be what the compiler needs to understand that the statement cannot be optimized out. I tested this with Clang 7 and Clang trunk (11) and they both behave this way.

File: bpf/lib/policy_log.h
Comment (pchaigno) [reply bpf/lib/policy_log.h]: I would define `POLICY_VERDICT_LOG_FILTER` at the top of this file if it isn't already defined. That should silence build warnings. Files that actually use `send_policy_verdict_notify()` (i.e., `bpf_lxc.c` right now) need to include `lxc_config.h` before including this file; I think that's already the case.

Comment (lzang) [reply bpf/lib/policy_log.h]: Thanks for the suggestion. Updated the patch.

**Comment by lzang** [comment]
test-me-please

File: bpf/lib/static_data.h
Comment (pchaigno) [inline bpf/lib/static_data.h]: All `DEFINE_xxx` macros now declare their variable as volatile, so there is no need anymore for a new macro :-)

Comment (lzang) [reply bpf/lib/static_data.h]: Thanks for pointing this out. Updated the patch.

**Comment by joestringer** [comment]
I filed #10615 for the Travis flake.

**Comment by lzang** [comment]
test-me-please

**Comment by lzang** [comment]
Thanks @joestringer ! For the "log buffer too small" issue, it is possible to put in the mitigation with iproute2 to the CI?

**Comment by pchaigno** [comment]
> Thanks @joestringer ! For the "log buffer too small" issue, it is possible to put in the mitigation with iproute2 to the CI?

That wouldn't solve the underlying issue, which is a complexity error in the BPF verifier (we're hitting the limit). As a short term solution, we're hoping #10542 will give us some slack there and allow the currently-blocked PRs to pass.

**Comment by lzang** [comment]
> That wouldn't solve the underlying issue, which is a complexity error in the BPF verifier (we're hitting the limit). As a short term solution, we're hoping #10542 will give us some slack there and allow the currently-blocked PRs to pass.

Yes, I know it doesn't solve the underlying issue, but I think it shouldn't block CI as it will take longer time for the real fix to get in. We can always remove the workaround later. It would great if #10542 can help and unblock CI.

**Comment by pchaigno** [comment]
Hi @lzang :wave: Could you rebase your pull request against master? #10542 was merged yesterday. It updates the LLVM/Clang version we use and should fix the CI issues you're currently hitting.

**Comment by lzang** [comment]
test-me-please

File: bpf/lib/policy_log.h
Comment (pchaigno) [inline bpf/lib/policy_log.h]: Looks like this is from a previous version.

File: bpf/lxc_config.h
Comment (pchaigno) [inline bpf/lxc_config.h:21]: Is this needed given its defined in `bpf/lib/policy_log.h` already? If you remove this, the `#ifndef POLICY_VERDICT_LOG_FILTER` guard in `bpf/lib/policy_log.h` can also be removed.

File: bpf/netdev_config.h
Comment (pchaigno) [inline bpf/netdev_config.h]: Ditto.

**Review by pchaigno — COMMENTED** [review]
Neat! The tests are passing :smiley: 

I have a couple small comments below, but we should be okay to merge after that :tada:

File: bpf/lib/policy_log.h
Comment (lzang) [reply bpf/lib/policy_log.h]: Thanks for the catch! Interesting that compiler didn't catch this error.

File: bpf/lxc_config.h
Comment (lzang) [reply bpf/lxc_config.h:21]: In the real case, these configuration are written from control plane, so it will still be in this file, right?

File: bpf/lib/policy_log.h
Comment (pchaigno) [reply bpf/lib/policy_log.h]: Yeah, that was weird to me too. We compile test with different macros in [`bpf/Makefile`](https://github.com/cilium/cilium/blob/0b5e263ff749c283481d0cc3b1b6e9165292b2c3/bpf/Makefile#L118-L133), but the `POLICY_VERDICT_NOTIFY` macro isn't part of those. It only guards these few lines, so I'm not sure it's worth it after all.

**Comment by tklauser** [comment]
test-me-please

File: bpf/lxc_config.h
Comment (pchaigno) [reply bpf/lxc_config.h:21]: Hm, that's a good point. I think that, as long as we move the `DEFINE_U32` in `policy_log.h` before the `#ifdef POLICY_VERDICT_NOTIFY` guard (to ensure it's always present), we should be fine with removing it from `lxc_config.h`. Then, because the symbol will always be in the endpoint's object file, we can also remove the write in `writeStaticData()`. Unless @joestringer prefers that we keep all such `DEFINE_U32`s in the `lxc_config.h` file?

Comment (joestringer) [reply bpf/lxc_config.h:21]: If it's going to be (over)written by `writeStaticData()` then I think it should be in lxc_config.h, that way it shows up in `/var/run/cilium/state/EPID/lxc_config.h` and can be retrospectively inspected.

Comment (joestringer) [reply bpf/lxc_config.h:21]: I agree that we should ensure that the symbol is unconditionally defined, otherwise you can hit weird upgrade issues when switching the feature on/off.

File: bpf/lib/policy_log.h
Comment (joestringer) [reply bpf/lib/policy_log.h]: FWIW ideally the bpf/Makefile default build will build the most complex program possible, so that we catch complexity issues with particular configurations in CI before users hit them.

If this filter is adding complexity (which it is, since we've had CI issues in that area), then we should ensure that the corresponding option is enabled by default in the `bpf/Makefile`.

File: bpf/lxc_config.h
Comment (pchaigno) [reply bpf/lxc_config.h:21]: Right, if we remove from `writeStaticData()` it's going to be hard to know the value during debugging :-/
Let's keep it as is then. Sorry for the noise @lzang!

File: bpf/lib/policy_log.h
Comment (pchaigno) [reply bpf/lib/policy_log.h]: It's currently enabled by default in `lxc_config.h`; that should be enough, right?

**Review by pchaigno — APPROVED** [review]

File: bpf/lib/policy_log.h
Comment (joestringer) [reply bpf/lib/policy_log.h]: I guess that CI is already catching these, but I was thinking of `test/bpf/check-complexity.sh` which loads the version built with the options from the Makefile. Would be good to fix this up either in this PR or as an immediate follow-up PR.

Comment (pchaigno) [reply bpf/lib/policy_log.h]: Ah, right. I wasn't thinking about `check-complexity.sh`. I can take care of that in a follow-up PR; I suspect `POLICY_VERDICT_NOTIFY` isn't the only missing option.

File: bpf/lxc_config.h
Comment (lzang) [reply bpf/lxc_config.h:21]: Not a problem at all. Thanks a lot for your reviewing!

File: bpf/lib/policy_log.h
Comment (joestringer) [reply bpf/lib/policy_log.h]: I looked again at this PR and in this case and it's already achieved, no further changes are necessary.

**Review by joestringer — APPROVED** [review]

**Pull request merged**
